src/services/refactors/inlineVariable.ts

```diff
@@ -0,0 +1,131 @@
+import {
+    CancellationToken,
+    Debug,
+    Diagnostics,
+    emptyArray,
+    Expression,
+    FindAllReferences,
+    getLocaleSpecificMessage,
+    getTokenAtPosition,
+    isVariableDeclaration,
+    isVariableDeclarationInVariableStatement,
+    mapDefined,
+    Node,
+    NoopCancellationToken,
+    Program,
+    refactor,
+    SourceFile,
+    textChanges,
+    VariableDeclaration,
+} from "../_namespaces/ts";
+import { RefactorErrorInfo, registerRefactor } from "../_namespaces/ts.refactor";
+
+const refactorName = "Inline variable";
+const refactorDescription = getLocaleSpecificMessage(Diagnostics.Inline_variable);
+
+const inlineVariableAction = {
+    name: refactorName,
+    description: refactorDescription,
+    kind: "refactor" // TODO: Not sure what the logic for defining these is.
+};
+
+interface InliningInfo {
+    references: Node[];
+    declaration: VariableDeclaration;
+    replacement: Expression;
+}
+
+registerRefactor(refactorName, {
+    kinds: [ inlineVariableAction.kind ],
+
+    getAvailableActions(context) {
+        const {
+            cancellationToken,
+            file,
+            program,
+            preferences,
+            startPosition
+        } = context;
+
+        const info = getInliningInfo(file, startPosition, program, cancellationToken);
+
+        if (refactor.isRefactorErrorInfo(info) && preferences.provideRefactorNotApplicableReason) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [{
+                    ...inlineVariableAction,
+                    notApplicableReason: info.error
+                }]
+            }];
+        }
+
+        if (info) {
+            return [{
+                name: refactorName,
+                description: refactorDescription,
+                actions: [inlineVariableAction]
+            }];
+        }
+
+        return emptyArray;
+    },
+
+    getEditsForAction(context, actionName) {
+        Debug.assert(actionName === refactorName, "Unexpected refactor invoked");
+
+        const {
+            cancellationToken,
+            file,
+            program,
+            startPosition
+        } = context;
+
+        const info = getInliningInfo(file, startPosition, program, cancellationToken);
+
+        if (!info || refactor.isRefactorErrorInfo(info)) {
+            return undefined;
+        }
+
+        const { references, declaration, replacement } = info;
+        const edits = textChanges.ChangeTracker.with(context, tracker => {
+            for (const node of references) {
+                tracker.replaceNode(file, node, replacement);
+            }
+            tracker.delete(file, declaration);
+        });
+
+        return { edits };
+    }
+});
+
+function getInliningInfo(file: SourceFile, startPosition: number, program: Program, cancellationToken: CancellationToken = NoopCancellationToken): InliningInfo | RefactorErrorInfo | undefined {

```

Comment:
Looks like you always are passing a `cancellationToken`?

```suggestion
function getInliningInfo(file: SourceFile, startPosition: number, program: Program, cancellationToken: CancellationToken): InliningInfo | RefactorErrorInfo | undefined {
```

Reply:
No because the `cancellationToken` from the refactor context can be undefined, so I have this fallback.